state-sync/state-sync-v2/state-sync-driver/src/continuous_syncer.rs

```diff
@@ -0,0 +1,381 @@
+    storage_synchronizer::StorageSynchronizerInterface, utils,
+};
+use data_streaming_service::{
+    data_notification::{DataPayload, NotificationId},
+    data_stream::DataStreamListener,
+    streaming_client::{DataStreamingClient, NotificationFeedback, StreamingServiceClient},
+};
+use diem_config::config::SyncingMode;
+use diem_infallible::Mutex;
+use diem_types::{
+    contract_event::ContractEvent,
+    epoch_change::Verifier,
+    ledger_info::LedgerInfoWithSignatures,
+    transaction::{
+        default_protocol::{TransactionListWithProof, TransactionOutputListWithProof},
+        Version,
+    },
+};
+use event_notifications::{EventNotificationSender, EventSubscriptionService};
+use std::sync::Arc;
+
+/// A simple component that manages the continuous syncing of the node
+pub struct ContinuousSyncer<S> {
+    // The currently active data stream (provided by the data streaming service)
+    active_data_stream: Option<DataStreamListener>,
+
+    // The config of the state sync driver
+    driver_configuration: DriverConfiguration,
+
+    // The event subscription service to notify listeners of on-chain events
+    event_subscription_service: Arc<Mutex<EventSubscriptionService>>,
+
+    // The client through which to stream data from the Diem network
+    streaming_service_client: StreamingServiceClient,
+
+    // The storage synchronizer used to update local storage
+    storage_synchronizer: Arc<Mutex<S>>,
+}
+
+impl<S: StorageSynchronizerInterface> ContinuousSyncer<S> {
+    pub fn new(
+        driver_configuration: DriverConfiguration,
+        event_subscription_service: Arc<Mutex<EventSubscriptionService>>,
+        streaming_service_client: StreamingServiceClient,
+        storage_synchronizer: Arc<Mutex<S>>,
+    ) -> Self {
+        Self {
+            active_data_stream: None,
+            driver_configuration,
+            event_subscription_service,
+            streaming_service_client,
+            storage_synchronizer,
+        }
+    }
+
+    /// Checks if the continuous syncer is able to make progress
+    pub async fn drive_progress(
+        &mut self,
+        consensus_sync_request: Arc<Mutex<Option<ConsensusSyncRequest>>>,
+    ) -> Result<(), Error> {
+        if self.active_data_stream.is_some() {
+            // We have an active data stream. Process any notifications!
+            self.process_active_stream_notifications(consensus_sync_request)
+                .await
+        } else {
+            // Fetch a new data stream to start streaming data
+            self.initialize_active_data_stream(consensus_sync_request)
+                .await
+        }
+    }
+
+    /// Initializes an active data stream so that we can begin to process notifications
+    async fn initialize_active_data_stream(
+        &mut self,
+        consensus_sync_request: Arc<Mutex<Option<ConsensusSyncRequest>>>,
+    ) -> Result<(), Error> {
+        // Fetch transactions or outputs starting at highest_synced_version + 1
```

Comment:
Do you want to check when initializing if there is already a data stream?

Reply:
This is only called by `drive_progress` above, which does exactly that 😄